src/container/srv_target.c

```diff
@@ -2537,3 +2657,114 @@ ds_cont_ec_timestamp_update(struct ds_cont_child *cont)
 {
 	cont->sc_ec_update_timestamp = d_hlc_get();
 }
+
+void
+ds_cont_child_insert_dtx(struct ds_cont_child *cont_child, struct dtx_leader_handle *new_dlh)
+{
+	struct dtx_leader_handle *dlh = NULL;
+
+	/* Open DTXs should be ordered in the list, and most likely the new inserted DTX
+	 * should be inserted at the head of the list.
+	 */
+	d_list_for_each_entry(dlh, &cont_child->sc_open_dtx_list, dlh_link_list) {
+		if (dlh->dlh_handle.dth_epoch < new_dlh->dlh_handle.dth_epoch) {
+			d_list_add_tail(&new_dlh->dlh_link_list, &dlh->dlh_link_list);
+			return;
+		}
+	}
+
+	d_list_add_tail(&new_dlh->dlh_link_list, &cont_child->sc_open_dtx_list);
+}
+
+static void
+cont_child_update_commit_eph(struct ds_cont_child *cont_child)
+{
+	struct dtx_leader_handle *dlh = NULL;
+
+	if (!cont_child->sc_dtx_commit_eph)
+		return;
+
+	/* First Get the minimum epoch from COS list */
+	*cont_child->sc_dtx_commit_eph = dtx_get_min_cos_eph(cont_child);
+
+	/* Then compare it with the open list, since the open list is naturally sorted,
+	 * so only compare the last one */
+	d_list_for_each_entry_reverse(dlh, &cont_child->sc_open_dtx_list, dlh_link_list) {
+		if (*cont_child->sc_dtx_commit_eph > dlh->dlh_handle.dth_epoch - 1)
+			*cont_child->sc_dtx_commit_eph = dlh->dlh_handle.dth_epoch - 1;
+		break;
+	}
+
+	D_DEBUG(DB_MD, DF_UUID" update commit eph to "DF_X64"\n", DP_UUID(cont_child->sc_uuid),
+		*cont_child->sc_dtx_commit_eph);
+}
+
+struct refresh_vos_agg_eph_arg {
+	uuid_t	pool_uuid;
+	uuid_t  cont_uuid;
+	daos_epoch_t min_eph;
+};
+
+int
+cont_refresh_vos_agg_eph_one(void *data)
+{
+	struct refresh_vos_agg_eph_arg *arg = data;
+	struct ds_cont_child	*cont_child;
+	int			rc;
+
+	rc = ds_cont_child_lookup(arg->pool_uuid, arg->cont_uuid, &cont_child);
+	if (rc)
+		return rc;
+
+	D_DEBUG(DB_MD, DF_CONT": %s agg boundary eph "DF_X64"->"DF_X64"\n",
+		DP_CONT(arg->pool_uuid, arg->cont_uuid),
+		cont_child->sc_ec_agg_eph_boundary < arg->min_eph ? "update" : "ignore",
+		cont_child->sc_ec_agg_eph_boundary, arg->min_eph);
+
+	if (cont_child->sc_ec_agg_eph_boundary < arg->min_eph) {
```

Comment:
just confirm how could it possible for this case? only for the case of exclude the target and reinit the target, and the min_eph move ahead between the exclude and reint?

Reply:
out of order of IV, NB: IV are asynchronous for this case.

Reply:
ah, sorry I was thought it as "sc_ec_agg_eph", sc_ec_agg_eph should never be < arg->min_eph